src/dmd/statementsem.d

```diff
@@ -733,9 +733,20 @@ package (dmd) extern (C++) final class StatementSemanticVisitor : Visitor
         {
             assert(oaggr.type);
 
-            fs.error("invalid `foreach` aggregate `%s` of type `%s`", oaggr.toChars(), oaggr.type.toPrettyChars());
-            if (isAggregate(fs.aggr.type))
-                fs.loc.errorSupplemental("maybe define `opApply()`, range primitives, or use `.tupleof`");
+            fs.error("invalid `%s` aggregate `%s` of type `%s`",
+                Token.toChars(fs.op), oaggr.toChars(), oaggr.type.toPrettyChars());
+
+            if (auto ad = isAggregate(fs.aggr.type))
+            {
+                if (fs.op == TOK.foreach_reverse_ && ad.search(Loc.initial, Id.Ffront))
+                {
+                    fs.loc.errorSupplemental(
+                        "`foreach_reverse` requires a bidirectional range (implementing `back` and `popBack`)");
```

Comment:
This is not an improvement. foreach_reverse does not require aggregates to be bidirectional ranges. It works with bidirectional_ranges, opApplyReverse, or the result of.tupleof. All the info is there in the original message, though it's wrong about `opApply`.  The problem is that it's poorly worded, so is probably only clear to someone who knows D pretty well.

So the message should be something like:

"`foreach_reverse` works with bidirectional ranges (implementing `back` and `popBack`), aggregates implementing `opApply`, or the result of an aggregate's `.tupleof` property"